language/move-vm/types/src/loaded_data/runtime_types.rs

```diff
@@ -14,6 +14,7 @@ pub struct StructType {
     pub fields: Vec<Type>,
     pub abilities: AbilitySet,
     pub type_parameters: Vec<AbilitySet>,
+    pub type_param_kinds: Vec<TypeParamKind>,
```

Comment:
I think we should have a discussion about whether to prefer [AoS (Array of Structs) or SoA (Struct of Arrays)](https://en.wikipedia.org/wiki/AoS_and_SoA) here. SoA is what the PR currently implements. Following is an example of AoS.
```
struct TypeParameter {
    is_phantom: bool,
    abilities: AbilitySet,
}

struct StructType {
    ..
    pub type_parameters: Vec<TypeParameter>,
    ..
}
```
AoS seems more intuitive and unlike SoA, we don't have to check that the two arrays have the same length. However one advantage SoA has is that we will continue to be able to pass the old list of ability sets to functions that expect `&[AbilitySet]`. Although with AoS, the issue can be fixed by defining a getter trait.
```
trait TypeParamGetAbilitySet {
    fn ability_set() -> &AbilitySet;
}

fn consumer(ty_params: &[AbilitySet]);
// changes to
fn consumer(ty_params: &[impl TypeParamGetAbilitySet]);
```

@tnowacki thoughts?

Reply:
We briefly discussed SoA with @tnowacki that's why I ended up implementing it. But after implementing it I think I lean towards AoS. Maintaining the invariant that both array are the same length can get very nasty and it's not very clear where to add the checks (I didn't find any other similar well-formed checks). The getter trait sounds like a good middle ground, didn't thought of that.

Reply:
You can do what you want here. It might make sense here to do `Vec<TypeParameterType>` or something. 

But in the file format I think it is important to keep them separate. See that discussion in the serializer.